### DISCUSSION TITLE
Comments: How to validate Express requests using io-ts

### DISCUSSION THREAD

**Discussion by jiayihu** [jiayihu/blog#63] [Announcements]
This issue is reserved for comments to [How to validate Express requests using io-ts](http://blog.jiayihu.net/how-to-validate-express-requests-using-io-ts/). Leave a comment below and it will be shown in the blog page.

**Comment by Nicky-P** [comment]
Awesome article, thanks for the upload! I'm having some problems following the "Transform io-ts error towards a more standard error response" section. Would you mind sharing what version of io-ts you are using?

**Comment by jiayihu** [comment]
Hi, I was using version `2.2.1`. That part may be indeed tricky, you want to consume a flat collection of errors instead of a tree. Let me know if you have any question

**Comment by Nicky-P** [comment]
Gotcha, thank you! Looks like some built in error handling was added. https://github.com/gcanti/io-ts/blob/master/Decoder.md#built-in-error-reporter 

I went ahead and used this approach with io-ts version 2.2.9

```
import { RequestHandler } from 'express';
import * as D from 'io-ts/lib/Decoder';
import { Decoder } from 'io-ts/lib/Decoder';
import { pipe } from 'fp-ts/lib/pipeable';
import { fold } from 'fp-ts/lib/Either';
import { ParamsDictionary } from 'express-serve-static-core';

export const validator: <T, A>(decoder: Decoder<T, A>) => RequestHandler<ParamsDictionary, any, T> = (decoder) => (req, res, next) => {
  return pipe(
    decoder.decode(req.body),
    fold(
      (errors) => res.status(400).send({ code: 'BadArgument', status: 'error', error: D.draw(errors) }),
      () => next(),
    ),
  );
};
```

**Comment by jiayihu** [comment]
That's cool, I'm personally in the same Italian slack channel with Giulio but he deploys new releases so often that's hard even for me to follow all the updates :D 
Thanks for reporting, I may add a note later in the article

**Comment by ak5** [comment]
When I do

```ts
return addIsland(req.body) // req.body is strictly typed
```

TS still thinks `req.body` is `any` - is there something I have to do to express to get the type to stay in place outside the middleware? Right now I have to do something like

```ts
return addIsland(req.body as D.TypeOf<typeof IslandDec>) // req.body is strictly typed
```

Does your `addIsland` deal with this or am I missing something?

**Comment by jiayihu** [comment]
The `T` generic in `const validator: <T>(decoder: Decoder<T>) => RequestHandler<ParamsDictionary, any, T>` is what makes the decoded value strictly typed. In any case just check express types to understand how the generics are passed in the flow
